README.md
# Single C File Taint Path Annotate

This vscode extension uses [SVF Framework](https://github.com/SVF-tools/SVF) and according to the [teaching materials](https://github.com/SVF-tools/Teaching-Software-Analysis) to create the vscode extension which can highlight the taint path of single .c file.

## Features

> It reads SrcSnk.txt which include $source$ and $sink$, {cFilename}.c which is the source code that needs to be analyzed. It will automatically generates llvm ir file {cFilename}.ll and taint.txt which contains the taint path with SVF.

## Requirements

MacOS or Linux system installed llvm compiler. The extension will use `clang -S -emit-llvm -O0 -g ${sourceFile} -o ${outputIRFile}` to compile the `.c` file into llvm ir file, and then can be analyzed with SVF.

## Extension Settings

It currently supports two commands:

```json
{
    "command": "extension.highlightTaintedPathsText",
    "title": "Highlight Tainted Paths From txt File"
},
{
    "command": "extension.highlightTaintedPaths",
    "title": "Analyze and Then Highlight Tainted Paths From C File"
}
```

- **highlightTaintedPaths** works on the currently opened `.c` file, and it also requires a `SrcSnk.txt` file in the same directory with `.c` file. The extension should generate a `.ll` file which is llvm ir file, and `taint.txt` which contains the taint path of the  C code file.

  - The folder of your code to analyze:

      ```
      code folder/
      ├─ code_to_analyze.c
      ├─ SrcSnk.txt
      ```

  - After running the command, you folder should be:

    ```
    code folder/
    ├─ code_to_analyze.c
    ├─ SrcSnk.txt
    ├─ taint.txt
    ├─ code_to_analyze.ll
    ```
  The extension should generate two files, and annotate the code with `taint.txt`.

- **highlightTaintedPathsIR** annotate current opened `.c` file with *exsisted* `cFilename.txt` which contains same content with `taint.txt`. It won't perform code analyze and just simply annotate with given txt file. 
  
  - This command only require .c file and taint.txt
      ```
        code folder/
        ├─ code_to_analyze.c
        ├─ code_to_analyze.txt (same name with your code filename)
      ```
  - Taint path text file format:
     ```
     { ln: 9  cl: 25  fl: test4.c } -->
     { ln: 13  cl: 13  fl: test4.c } -->
     ```

## Known Issues

The bundled binary file only supports Unix like file path. It is created with MacOS, and soon to be tested on Ubuntu.

## Release Notes

### 1.0.0

Initial release of extension.

### 1.0.1

Fix bug: the highlight should be processed after the taint.txt file generates.


CHANGELOG.md
# Change Log

All notable changes to the "test-annotate" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';
import * as cp from 'child_process';
import { spawn } from 'child_process';
import { log } from 'console';

function compileToLLVMIR(sourceFile: string, outputIRFile: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const clangCommand = `clang -S -emit-llvm -O0 -g ${sourceFile} -o ${outputIRFile}`;

    exec(clangCommand, (error, stdout, stderr) => {
      if (error) {
        reject(error);
        return;
      }
      if (stderr) {
        reject(new Error(stderr));
        return;
      }
      resolve();
    });
  });
}

function runBinaryProgram(programPath: string, args: string[]): Promise<void> {
  return new Promise((resolve, reject) => {
    const childProcess = spawn(programPath, args);

    childProcess.stdout.on('data', (data) => {
      // Output the data to the VSCode output channel
      vscode.window.activeTerminal?.sendText(data.toString());
    });

    childProcess.stderr.on('data', (data) => {
      // Output the error to the VSCode output channel
      vscode.window.activeTerminal?.sendText(data.toString());
    });

    childProcess.on('error', (error) => {
      // Type assertion to specify the type of 'error' as 'Error'
      const err: Error = error as Error;
      // Reject the promise with the error
      reject(err);
    });

    childProcess.on('close', (code) => {
      if (code === 0) {
        // Resolve the promise when the process completes successfully
        resolve();
      } else {
        // Reject the promise if the process exits with a non-zero code
        reject(new Error(`Process exited with code ${code}`));
      }
    });
  });
}

function convertTextToJson(text: string): any {
  const regex = /{ ln:\s+(\d+)\s+cl: (\d+)\s+fl: (.+) }/g;
  const matches = regex.exec(text);

  if (matches) {
    const [, lineNumber, columnNumber, filePath] = matches;
    return {
      "ln": parseInt(lineNumber),
      "cl": parseInt(columnNumber),
      "fl": filePath
    };
  }

  return null;
}

function removeFileExtension(filePath: string): string {
  const fileExtension = path.extname(filePath);
  const fileNameWithoutExtension = path.basename(filePath, fileExtension);
  const directory = path.dirname(filePath);
  return path.join(directory, fileNameWithoutExtension);
}

function getPreviousSpaceOrNewlinePosition(text: string, startPosition: number): number {
  const substring = text.substring(startPosition);
  const spaceIndex = substring.indexOf(' ');
  const newlineIndex = substring.indexOf('\n');

  if (spaceIndex === -1 && newlineIndex === -1) {
      // If neither space nor newline is found, return -1
      return -1;
  }

  if (spaceIndex === -1) {
      // If only newline is found, return its index
      return startPosition + newlineIndex;
  }

  if (newlineIndex === -1) {
      // If only space is found, return its index
      return startPosition + spaceIndex;
  }

  // Return the index of the first occurrence of either space or newline
  return startPosition + Math.min(spaceIndex, newlineIndex);
}

function getLineLength(filePath: string, lineNumber: number, columnNumber: number): number {
  try {
    const fileContent = fs.readFileSync(filePath, 'utf-8');
    const lines = fileContent.split('\n');
    const line = lines[lineNumber - 1];
    const posi = getPreviousSpaceOrNewlinePosition(line, columnNumber);
    if (posi !== -1) {
      return posi;
    }
    return line.length;
  } catch (error) {
    console.error('Error occurred while reading the file:', error);
    return 0;
  }
}

async function parseTaintFlowFile(taintFile: string, cFile: string): Promise<vscode.DecorationOptions[]> {
  const taintPaths: vscode.DecorationOptions[] = [];
  const taintFlowFilePath = `${taintFile}`;

  if (!fs.existsSync(taintFlowFilePath)) {
    console.error(`Taint flow file not found: ${taintFlowFilePath}`);
    return taintPaths;
  }

  const taintFlowData = fs.readFileSync(taintFlowFilePath, 'utf-8');

  const pathSegments = taintFlowData.split('-->');
  for (let i = 0; i < pathSegments.length; i++) {
    const segment = pathSegments[i].trim();
    if (segment.startsWith('{') && segment.endsWith('}')) {
      try {
        const newsegment = convertTextToJson(segment);
        // const taintInfo = JSON.parse(newsegment);

        const { ln, cl, fl } = newsegment;

        const linelen = await getLineLength(cFile, ln, cl);
        if (linelen === -1) {
          console.error(`Line ${cl} does not exist in the file or an error occurred while reading.`);
          continue;
        }

        const decoration: vscode.DecorationOptions = {
          range: new vscode.Range(ln - 1, cl - 1, ln - 1, linelen),
          hoverMessage: `Tainted Path Segment: ${fl}:${ln}:${cl}`,
        };

        taintPaths.push(decoration);
      } catch (error) {
        console.error('Error parsing taint flow data:', error);
      }
    }
  }

  return taintPaths;
}

export const sleep = (ms: number)=> {
  return new Promise(resolve=>setTimeout(resolve, ms));
};

export function activate(context: vscode.ExtensionContext) {
  let disposable = vscode.commands.registerCommand('extension.highlightTaintedPaths', async () => {
    try {
      const activeEditor = vscode.window.activeTextEditor;
      if (activeEditor) {
        const cFilePath = activeEditor.document.uri.fsPath;
        const currentFileNameWithoutExtension = removeFileExtension(cFilePath);
        const llFilePath = `${currentFileNameWithoutExtension}.ll`;

        const currentFileDirectory = path.dirname(__filename);
        const programPath = path.join(currentFileDirectory, 'analyze');
        const args = [llFilePath];

        compileToLLVMIR(cFilePath, llFilePath)
          .then(() => {
            console.log('Compilation successful!');
            runBinaryProgram(programPath, args)
              .then(() => {
                vscode.window.showInformationMessage('Binary program execution completed successfully!');
                const taintFlowFilePath = path.join(path.dirname(cFilePath), 'taint.txt');
                parseTaintFlowFile(taintFlowFilePath, cFilePath)
                  .then((taintPaths) => {
                    const decorationType = vscode.window.createTextEditorDecorationType({
                      backgroundColor: 'rgba(127,255,127,0.07)',
                      border: '2px solid white',
                      fontWeight: 'bold',
                    });

                    activeEditor.setDecorations(decorationType, taintPaths);
                  })
                  .catch((error) => {
                    vscode.window.showErrorMessage(`Taint flow file parsing failed: ${error}`);
                  });
              })
              .catch((error) => {
                vscode.window.showErrorMessage(`Binary program execution failed: ${error}`);
              });
          })
          .catch((error) => {
            console.error('Compilation failed:', error);
          });
      }
    } catch (error) {
      console.error(error);
    }
  });

  let disposable2 = vscode.commands.registerCommand('extension.highlightTaintedPathsText', async () => {
    try {
      const activeEditor = vscode.window.activeTextEditor;
      if (activeEditor) {
        const cFilePath = activeEditor.document.uri.fsPath;
        const currentFileNameWithoutExtension = removeFileExtension(cFilePath);

        // const taintFlowFilePath = `${currentFileNameWithoutExtension}.txt`;
        const taintFlowFilePath = `${currentFileNameWithoutExtension}.txt`;

        if (fs.existsSync(taintFlowFilePath)) {
          const taintPaths = await parseTaintFlowFile(taintFlowFilePath, cFilePath);

          const decorationType = vscode.window.createTextEditorDecorationType({
            backgroundColor: 'rgba(255, 0, 0, 0.3)',
          });

          activeEditor.setDecorations(decorationType, taintPaths);
        } else {
          vscode.window.showErrorMessage(`Taint flow file not found: ${taintFlowFilePath}`);
        }
      }
    } catch (error) {
      console.error(error);
    }
  });

  context.subscriptions.push(disposable);
  context.subscriptions.push(disposable2);
}

export function deactivate() { }


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = exports.sleep = void 0;
const vscode = require("vscode");
const fs = require("fs");
const path = require("path");
const child_process_1 = require("child_process");
const child_process_2 = require("child_process");
function compileToLLVMIR(sourceFile, outputIRFile) {
    return new Promise((resolve, reject) => {
        const clangCommand = `clang -S -emit-llvm -O0 -g ${sourceFile} -o ${outputIRFile}`;
        (0, child_process_1.exec)(clangCommand, (error, stdout, stderr) => {
            if (error) {
                reject(error);
                return;
            }
            if (stderr) {
                reject(new Error(stderr));
                return;
            }
            resolve();
        });
    });
}
function runBinaryProgram(programPath, args) {
    return new Promise((resolve, reject) => {
        const childProcess = (0, child_process_2.spawn)(programPath, args);
        childProcess.stdout.on('data', (data) => {
            // Output the data to the VSCode output channel
            vscode.window.activeTerminal?.sendText(data.toString());
        });
        childProcess.stderr.on('data', (data) => {
            // Output the error to the VSCode output channel
            vscode.window.activeTerminal?.sendText(data.toString());
        });
        childProcess.on('error', (error) => {
            // Type assertion to specify the type of 'error' as 'Error'
            const err = error;
            // Reject the promise with the error
            reject(err);
        });
        childProcess.on('close', (code) => {
            if (code === 0) {
                // Resolve the promise when the process completes successfully
                resolve();
            }
            else {
                // Reject the promise if the process exits with a non-zero code
                reject(new Error(`Process exited with code ${code}`));
            }
        });
    });
}
function convertTextToJson(text) {
    const regex = /{ ln:\s+(\d+)\s+cl: (\d+)\s+fl: (.+) }/g;
    const matches = regex.exec(text);
    if (matches) {
        const [, lineNumber, columnNumber, filePath] = matches;
        return {
            "ln": parseInt(lineNumber),
            "cl": parseInt(columnNumber),
            "fl": filePath
        };
    }
    return null;
}
function removeFileExtension(filePath) {
    const fileExtension = path.extname(filePath);
    const fileNameWithoutExtension = path.basename(filePath, fileExtension);
    const directory = path.dirname(filePath);
    return path.join(directory, fileNameWithoutExtension);
}
function getPreviousSpaceOrNewlinePosition(text, startPosition) {
    const substring = text.substring(startPosition);
    const spaceIndex = substring.indexOf(' ');
    const newlineIndex = substring.indexOf('\n');
    if (spaceIndex === -1 && newlineIndex === -1) {
        // If neither space nor newline is found, return -1
        return -1;
    }
    if (spaceIndex === -1) {
        // If only newline is found, return its index
        return startPosition + newlineIndex;
    }
    if (newlineIndex === -1) {
        // If only space is found, return its index
        return startPosition + spaceIndex;
    }
    // Return the index of the first occurrence of either space or newline
    return startPosition + Math.min(spaceIndex, newlineIndex);
}
function getLineLength(filePath, lineNumber, columnNumber) {
    try {
        const fileContent = fs.readFileSync(filePath, 'utf-8');
        const lines = fileContent.split('\n');
        const line = lines[lineNumber - 1];
        const posi = getPreviousSpaceOrNewlinePosition(line, columnNumber);
        if (posi !== -1) {
            return posi;
        }
        return line.length;
    }
    catch (error) {
        console.error('Error occurred while reading the file:', error);
        return 0;
    }
}
async function parseTaintFlowFile(taintFile, cFile) {
    const taintPaths = [];
    const taintFlowFilePath = `${taintFile}`;
    if (!fs.existsSync(taintFlowFilePath)) {
        console.error(`Taint flow file not found: ${taintFlowFilePath}`);
        return taintPaths;
    }
    const taintFlowData = fs.readFileSync(taintFlowFilePath, 'utf-8');
    const pathSegments = taintFlowData.split('-->');
    for (let i = 0; i < pathSegments.length; i++) {
        const segment = pathSegments[i].trim();
        if (segment.startsWith('{') && segment.endsWith('}')) {
            try {
                const newsegment = convertTextToJson(segment);
                // const taintInfo = JSON.parse(newsegment);
                const { ln, cl, fl } = newsegment;
                const linelen = await getLineLength(cFile, ln, cl);
                if (linelen === -1) {
                    console.error(`Line ${cl} does not exist in the file or an error occurred while reading.`);
                    continue;
                }
                const decoration = {
                    range: new vscode.Range(ln - 1, cl - 1, ln - 1, linelen),
                    hoverMessage: `Tainted Path Segment: ${fl}:${ln}:${cl}`,
                };
                taintPaths.push(decoration);
            }
            catch (error) {
                console.error('Error parsing taint flow data:', error);
            }
        }
    }
    return taintPaths;
}
const sleep = (ms) => {
    return new Promise(resolve => setTimeout(resolve, ms));
};
exports.sleep = sleep;
function activate(context) {
    let disposable = vscode.commands.registerCommand('extension.highlightTaintedPaths', async () => {
        try {
            const activeEditor = vscode.window.activeTextEditor;
            if (activeEditor) {
                const cFilePath = activeEditor.document.uri.fsPath;
                const currentFileNameWithoutExtension = removeFileExtension(cFilePath);
                const llFilePath = `${currentFileNameWithoutExtension}.ll`;
                const currentFileDirectory = path.dirname(__filename);
                const programPath = path.join(currentFileDirectory, 'analyze');
                const args = [llFilePath];
                compileToLLVMIR(cFilePath, llFilePath)
                    .then(() => {
                    console.log('Compilation successful!');
                    runBinaryProgram(programPath, args)
                        .then(() => {
                        vscode.window.showInformationMessage('Binary program execution completed successfully!');
                        const taintFlowFilePath = path.join(path.dirname(cFilePath), 'taint.txt');
                        parseTaintFlowFile(taintFlowFilePath, cFilePath)
                            .then((taintPaths) => {
                            const decorationType = vscode.window.createTextEditorDecorationType({
                                backgroundColor: 'rgba(127,255,127,0.07)',
                                border: '2px solid white',
                                fontWeight: 'bold',
                            });
                            activeEditor.setDecorations(decorationType, taintPaths);
                        })
                            .catch((error) => {
                            vscode.window.showErrorMessage(`Taint flow file parsing failed: ${error}`);
                        });
                    })
                        .catch((error) => {
                        vscode.window.showErrorMessage(`Binary program execution failed: ${error}`);
                    });
                })
                    .catch((error) => {
                    console.error('Compilation failed:', error);
                });
            }
        }
        catch (error) {
            console.error(error);
        }
    });
    let disposable2 = vscode.commands.registerCommand('extension.highlightTaintedPathsText', async () => {
        try {
            const activeEditor = vscode.window.activeTextEditor;
            if (activeEditor) {
                const cFilePath = activeEditor.document.uri.fsPath;
                const currentFileNameWithoutExtension = removeFileExtension(cFilePath);
                // const taintFlowFilePath = `${currentFileNameWithoutExtension}.txt`;
                const taintFlowFilePath = `${currentFileNameWithoutExtension}.txt`;
                if (fs.existsSync(taintFlowFilePath)) {
                    const taintPaths = await parseTaintFlowFile(taintFlowFilePath, cFilePath);
                    const decorationType = vscode.window.createTextEditorDecorationType({
                        backgroundColor: 'rgba(255, 0, 0, 0.3)',
                    });
                    activeEditor.setDecorations(decorationType, taintPaths);
                }
                else {
                    vscode.window.showErrorMessage(`Taint flow file not found: ${taintFlowFilePath}`);
                }
            }
        }
        catch (error) {
            console.error(error);
        }
    });
    context.subscriptions.push(disposable);
    context.subscriptions.push(disposable2);
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map